src/main/java/org/opensearch/knn/index/codec/derivedsource/DerivedSourceVectorInjector.java

```diff
@@ -0,0 +1,132 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * This class is responsible for injecting vectors into the source of a document. From a high level, it uses alternative
+ *  format readers and information about the fields to inject vectors into the source.
+ */
+@Log4j2
+public class DerivedSourceVectorInjector implements Closeable {
+
+    private final DerivedSourceReaders derivedSourceReaders;
+    private final List<PerFieldDerivedVectorInjector> perFieldDerivedVectorInjectors;
+    private final Set<String> fieldNames;
+
+    /**
+     * Constructor for DerivedSourceVectorInjector.
+     *
+     * @param derivedSourceReadersSupplier Supplier for the derived source readers.
+     * @param segmentReadState Segment read state
+     * @param fieldsToInjectVector List of fields to inject vectors into
+     */
+    public DerivedSourceVectorInjector(
+        DerivedSourceReadersSupplier derivedSourceReadersSupplier,
+        SegmentReadState segmentReadState,
+        List<FieldInfo> fieldsToInjectVector
+    ) throws IOException {
+        this.derivedSourceReaders = derivedSourceReadersSupplier.getReaders(segmentReadState);
+        this.perFieldDerivedVectorInjectors = new ArrayList<>();
+        this.fieldNames = new HashSet<>();
+        for (FieldInfo fieldInfo : fieldsToInjectVector) {
+            this.perFieldDerivedVectorInjectors.add(
+                PerFieldDerivedVectorInjectorFactory.create(fieldInfo, derivedSourceReaders, segmentReadState)
+            );
+            this.fieldNames.add(fieldInfo.name);
+        }
+    }
+
+    /**
+     * Given a docId and the source of that doc as bytes, add all the necessary vector fields into the source.
+     *
+     * @param docId doc id of the document
+     * @param sourceAsBytes source of document as bytes
+     * @return byte array of the source with the vector fields added
+     * @throws IOException if there is an issue reading from the formats
+     */
+    public byte[] injectVectors(int docId, byte[] sourceAsBytes) throws IOException {
+        // Deserialize the source into a modifiable map
+        Tuple<? extends MediaType, Map<String, Object>> mapTuple = XContentHelper.convertToMap(
+            BytesReference.fromByteBuffer(ByteBuffer.wrap(sourceAsBytes)),
+            true,
+            MediaTypeRegistry.getDefaultMediaType()
+        );
+        // Have to create a copy of the map here to ensure that is mutable
+        Map<String, Object> sourceAsMap = new HashMap<>(mapTuple.v2());
+
+        // For each vector field, add in the source. The per field injectors are responsible for skipping if
+        // the field is not present.
+        for (PerFieldDerivedVectorInjector vectorInjector : perFieldDerivedVectorInjectors) {
+            vectorInjector.inject(docId, sourceAsMap);
+        }
+
+        // At this point, we can serialize the modified source map
+        BytesStreamOutput bStream = new BytesStreamOutput(1024);
```

Comment:
any reason for having this constant of 1024?

Reply:
Let me link code in core where this was copied from.